### DISCUSSION TITLE
Automatic gitlab ssh management

### DISCUSSION THREAD

**Discussion by dvv** [kanidm/kanidm#2500] [Q&A] [answered]
Hi!

I'm testing gitlab ce integration. It perfectly logs users in. Knowing that kanidm can provide ssh keys I would like to make gitlab honor those keys. For that I'm trying to tweak gitlab's sshd after documentation:
```
PubkeyAuthentication yes
AuthorizedKeysCommand /usr/sbin/kanidm_ssh_authorizedkeys_direct -H https://idm.example.com -D anonymous %u
AuthorizedKeysCommandUser nobody
UsePAM yes
```

Yet no success. I wonder if the above change to sshd config is sufficient or should I install and properly run `kanidm-unixd` inside gitlab container as well?

TIA

**Comment by Firstyear** [comment]
It depends a lot on how gitlab SSH is configured - it could simply be "not picking up the config". It may have it's own sshd config somewhere. So that may require some work to take apart.

You don't need to run unixd in the container - you can actually just expose the container hosts unixd AF_UNIX socket into the container and it will resolve via the host :)

**Comment by dvv** [comment]
Right, I found sshd config is read from custom place.
Now the situation is:

- if I do `ssh -T kanidm-user@gitlab` it chokes at sshd before `kanidm_ssh_authorizedkeys_direct` has a chance with "no such user" (indeed, these is only "git" user there).
- if I do `ssh -T git@gitlab` it chokes at kanidm with "no such user" (indeed, no "git" user at kanidm). And I don't know how to pass the kanidm user along with ssh (and hence git push)

I believe that is where kanidm pam would help.

**Reply by Firstyear** [reply, marked as answer]
Then you should be able to add the kani pam and nss modules to the container - as mentioned you can use the pam/nss/ssh tools via the bind mounted AF_UNIX socket from the hosts unixd, so you don't need to run the resolver directly in the container itself.

I think the only "gotcha" here is don't most git clients expect "git@host" where git must be the username? I didn't think it allowed non-standard usernames because there is some magic with how it handles the shell and permissions. So in this case we may need a different way to retrieve the ssh keys en-masse for every user ... which might be a bit much 🤣

**Reply by dvv** [reply]
Yeah, this would lead to nowhere probably

**Comment by dvv** [comment]
Natively it uses
```
AuthorizedKeysCommand /opt/gitlab/embedded/service/gitlab-shell/bin/gitlab-shell-authorized-keys-check git %u %k
```
As per https://man.openbsd.org/sshd_config#TOKENS `%k` is "The base64-encoded key or certificate for authentication".
I wonder if there's a way to mimic it with kanidm interaction: "given the base64 of a key, what is the preferred_username of the owner?"

**Reply by Firstyear** [reply]
Let's open an issue for it - searching by key should be very achievable, so we should be able to do this without a huge fuss. 

If you give that command a valid user and key, what output does it return? Does it return the ssh key, or some other formatted data?

**Reply by dvv** [reply]
Am exploring ^^^

**Reply by Firstyear** [reply]
Sounds good mate - thanks for your great discussions and issues, you've already helped make Kanidm a better project!

**Comment by dvv** [comment]
req:
```
/opt/gitlab/embedded/service/gitlab-shell/bin/gitlab-shell-authorized-keys-check git git AAAAC3NzaC1lZDI1NTE5AAAAIPSCjr5JScHoX304vv1iceNa+R5im5gItjlwDmnWllV9
```
res:
```
command="/opt/gitlab/embedded/service/gitlab-shell/bin/gitlab-shell key-2",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPSCjr5JScHoX304vv1iceNa+R5im5gItjlwDmnWllV9 Vladimir Dronnikov (test-gitlab.example.com)
```
`git git` is pretty dummy here I believe.
Then there goes the body of the key.

It outputs some crap then the key verbatim then "displayname".

Strange, but spoofing by echo-ing this ^^^ line still doesn't allow the user to enter.
Nevertheless, something like `kanidm_ssh_lookup_authorizedkey KEY-BODY` returning the preferred_username would allow to craft the above answer by calling `/usr/sbin/kanidm_ssh_authorizedkeys_direct -H https://idm.example.com -D anonymous THAT-USERNAME`.

**Reply by Firstyear** [reply]
Okay, I think this should be a feature request issue now, with these examples. We likely need a way to return extra info based on what was requested here.
